In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)
pd.set_option("display.float_format", "{:.2f}".format)

In [4]:
file_path = r"C:\Users\amit6\OneDrive\Desktop\Intelligent-Retail-Customer-Analytics\Data\raw_data\online_retail_II.xlsx"

excel_file = pd.ExcelFile(file_path)

excel_file.sheet_names

['Year 2009-2010', 'Year 2010-2011']

In [5]:
df_1 = pd.read_excel(file_path , sheet_name= 'Year 2009-2010')
df_2 = pd.read_excel(file_path , sheet_name= 'Year 2010-2011')
df_1.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.00,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.00,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.00,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.00,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.00,United Kingdom


In [7]:
df_2.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.00,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.00,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.00,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.00,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.00,United Kingdom


In [8]:
print(df_1.columns.tolist())
print(df_2.columns.tolist())

['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']
['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']


In [9]:
df_1.columns.equals(df_2.columns)

True

In [10]:
missing = pd.DataFrame({
    "2009": df_1.isnull().sum(),
    "2010": df_2.isnull().sum()
})

missing

,2009,2010
Invoice,0,0
StockCode,0,0
Description,2928,1454
Quantity,0,0
InvoiceDate,0,0
Price,0,0
Customer ID,107927,135080
Country,0,0


In [ ]:
print("2009 Sheet")
print(df_1["InvoiceDate"].min())
print(df_1["InvoiceDate"].max())

print("\n")

print("2010 Sheet")
print(df_2["InvoiceDate"].min())
print(df_2["InvoiceDate"].max())

## Data for 01 to 09 december in 2010 overlapes 

2009 Sheet
2009-12-01 07:45:00
2010-12-09 20:01:00


2010 Sheet
2010-12-01 08:26:00
2011-12-09 12:50:00


In [12]:
df_1["Invoice"].nunique()

28816

In [13]:
df_2["Invoice"].nunique()

25900

In [14]:
df_1["Customer ID"].nunique()

4383

In [15]:
df_2["Customer ID"].nunique()

4372

In [17]:
## Common Invoices present
common_invoice = set(df_1["Invoice"]).intersection(
    set(df_2["Invoice"])
)

len(common_invoice)

1088

In [18]:
## Common customer present
common_customer = set(df_1["Customer ID"]).intersection(
    set(df_2["Customer ID"])
)

len(common_customer)

2813

In [19]:
## Schema Validation
schema_valid = (
    df_1.columns.equals(df_2.columns)
    and
    all(df_1.dtypes == df_2.dtypes)
)

schema_valid

True

In [21]:
print(len(common_invoice))
list(common_invoice)[:20]

1088


[536576,
 536577,
 536578,
 536579,
 536580,
 536581,
 536582,
 536583,
 536584,
 536585,
 536586,
 536587,
 536588,
 536589,
 536590,
 536591,
 536592,
 536593,
 536594,
 536595]

In [ ]:
sample_invoice = list(common_invoice)[0]

print("2009 Sheet")
print(df_1[df_1["Invoice"] == sample_invoice])



2009 Sheet
       Invoice StockCode                          Description  Quantity  \
505247  536576     22041         RECORD FRAME 7" SINGLE SIZE        144   
505248  536576     85174                    S/4 CACTI CANDLES        16   
505249  536576     84050      PINK HEART SHAPE EGG FRYING PAN        96   
505250  536576    84029E       RED WOOLLY HOTTIE WHITE HEART.        96   
505251  536576     82615  PINK MARSHMALLOW SCARF KNITTING KIT        36   
505252  536576     21156            RETROSPOT CHILDRENS APRON        12   
505253  536576     20727              LUNCH BAG  BLACK SKULL.        60   
505254  536576    35004C      SET OF 3 COLOURED  FLYING DUCKS       120   
505255  536576    85099B              JUMBO BAG RED RETROSPOT        30   
505256  536576    85123A   WHITE HANGING HEART T-LIGHT HOLDER       128   
505257  536576     22095                 LADS ONLY TISSUE BOX        72   
505258  536576     85038      6 CHOCOLATE LOVE HEART T-LIGHTS        72   
505259  536576

In [23]:
print("\n2010 Sheet")
print(df_2[df_2["Invoice"] == sample_invoice])


2010 Sheet
     Invoice StockCode                          Description  Quantity  \
2309  536576     22041         RECORD FRAME 7" SINGLE SIZE        144   
2310  536576     85174                    S/4 CACTI CANDLES        16   
2311  536576     84050      PINK HEART SHAPE EGG FRYING PAN        96   
2312  536576    84029E       RED WOOLLY HOTTIE WHITE HEART.        96   
2313  536576     82615  PINK MARSHMALLOW SCARF KNITTING KIT        36   
2314  536576     21156            RETROSPOT CHILDRENS APRON        12   
2315  536576     20727              LUNCH BAG  BLACK SKULL.        60   
2316  536576    35004C      SET OF 3 COLOURED  FLYING DUCKS       120   
2317  536576    85099B              JUMBO BAG RED RETROSPOT        30   
2318  536576    85123A   WHITE HANGING HEART T-LIGHT HOLDER       128   
2319  536576     22095                 LADS ONLY TISSUE BOX        72   
2320  536576     85038      6 CHOCOLATE LOVE HEART T-LIGHTS        72   
2321  536576    15056N            EDWAR

In [28]:
last_date = df_1["InvoiceDate"].max()

df_2_modified = df_2[df_2["InvoiceDate"] > last_date]
df_2_modified

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
22523,538172,21562,HAWAIIAN GRASS SKIRT,12,2010-12-10 09:33:00,1.25,15805.00,United Kingdom
22524,538172,79321,CHILLI LIGHTS,8,2010-12-10 09:33:00,4.95,15805.00,United Kingdom
22525,538172,22041,"RECORD FRAME 7"" SINGLE SIZE",12,2010-12-10 09:33:00,2.55,15805.00,United Kingdom
22526,538172,84558A,3D DOG PICTURE PLAYING CARDS,12,2010-12-10 09:33:00,2.95,15805.00,United Kingdom
22527,538172,22952,60 CAKE CASES VINTAGE CHRISTMAS,24,2010-12-10 09:33:00,0.55,15805.00,United Kingdom
...,...,...,...,...,...,...,...,...
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.00,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.00,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.00,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.00,France


In [32]:
retail_df = pd.concat([df_1,df_2_modified],ignore_index=True)
retail_df

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.00,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.00,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.00,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.00,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.00,United Kingdom
...,...,...,...,...,...,...,...,...
1044843,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.00,France
1044844,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.00,France
1044845,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.00,France
1044846,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.00,France


In [33]:
# retail_df = retail_df.drop_duplicates()
print("Shape before removing duplicates:", retail_df.shape)

duplicates = retail_df.duplicated().sum()
print("Duplicate rows:", duplicates)

retail_df = retail_df.drop_duplicates()

print("Shape after removing duplicates:", retail_df.shape)

Shape before removing duplicates: (1044848, 8)
Duplicate rows: 11812
Shape after removing duplicates: (1033036, 8)


In [31]:
retail_df

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.00,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.00,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.00,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.00,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.00,United Kingdom
...,...,...,...,...,...,...,...,...
1044843,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.00,France
1044844,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.00,France
1044845,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.00,France
1044846,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.00,France


In [34]:
retail_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1033036 entries, 0 to 1044847
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1033036 non-null  object        
 1   StockCode    1033036 non-null  object        
 2   Description  1028761 non-null  object        
 3   Quantity     1033036 non-null  int64         
 4   InvoiceDate  1033036 non-null  datetime64[ns]
 5   Price        1033036 non-null  float64       
 6   Customer ID  797885 non-null   float64       
 7   Country      1033036 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 70.9+ MB


In [36]:
retail_df.to_csv("result_retail.csv", index=False)

### Duplicate Record Validation

After resolving the overlap between the two worksheets, the merged dataset was checked for exact duplicate rows.

- Total records before duplicate removal: **1,044,848**
- Exact duplicate rows identified: **11,812**
- Total records after duplicate removal: **1,033,036**

These duplicate records were removed to prevent double-counting of transactions in subsequent customer, sales, and forecasting analyses.